### PR TITLE
Fixes #12865: Vault plugin : Add a variable_from_vault generic method.

### PR DESCRIPTION
**Pull request by VictorQrt**
https://www.rudder-project.org/redmine/issues/12865

### CODE REVIEW

**Comment by VictorQrt** [comment]
Commit modified

File: vault/packaging/postinst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+cp -a /opt/rudder/share/plugins/vault/usr/share/ncf/tree /usr/share/ncf/
```
Comment (amousset) [inline vault/packaging/postinst]: @ncharles do you think we should put it with CFEgine stubs for dsc methods instead?

File: vault/share/plugins/vault/usr/share/ncf/tree/20_cfe_basics/vault_lib.cf [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      "vault.server_addr" string => "${vault.config[server_addr]}";
+
+    pass2.mode_userpass.vault_reachable::
+      "vault_apicall_output" string => execresult("/usr/bin/curl -s --request POST --data '{\"password\": \"${vault.config[auth][userpass][password]}\"}' ${vault.server_addr}/v1/auth/userpass/login/${vault.config[auth][userpass][user]}", "noshell");
+      
+      "data_output" data => "${vault_apicall_output}";
+
+    pass2.vault_reachable::
+      "vault.auth_token" string => "${data_output[auth][client_token]}",
+        ifvarclass => isvariable("data_output[auth][client_token]");
```
Comment (amousset) [inline vault/share/plugins/vault/usr/share/ncf/tree/20_cfe_basics/vault_lib.cf]: We should display an error message if the config is not defined

**Comment by VictorQrt** [comment]
Commit modified

File: vault/packaging/postinst [hunk 3]
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+mkdir -p /var/rudder/plugin-resources
+cp -a /opt/rudder/share/plugins/vault/* /var/rudder/configuration-repository/ncf/
+cd /var/rudder/configuration-repository/ncf/
+rm sample_vault.json
+git add . && git commit -m "Vault plugin installation"
```
Comment (ncharles) [inline vault/packaging/postinst]: You should probably mention somewhere in the output that there is a sample_vault.json

File: vault/share/plugins/vault/20_cfe_basics/vault_lib.cf [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#####################################################################################
+
+bundle agent vault_fetch_config()
```
Comment (ncharles) [inline vault/share/plugins/vault/20_cfe_basics/vault_lib.cf]: could you document here the format of the vault.json file ?

File: vault/share/plugins/vault/30_generic_methods/variable_from_vault.cf [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+#
+#####################################################################################
+#
+# @name Variable from Vault
+# @description Gets a key-value dictionnary from Vault given the secret path
+#
+# @documentation To use the generated variable, you must use the form `${variable_prefix.variable_name}` with each name replaced with the parameters of this method.
+#
+# Access to the vault has to be configured on each agent in /var/rudder/plugin-resources/vault.json. A sample config file is provided in /opt/rudder/share/plugins/vault/sample_vault.json
+#
```
Comment (ncharles) [inline vault/share/plugins/vault/30_generic_methods/variable_from_vault.cf]: can you specify the success & error case ? that fetching a key is a repair, failing is an error?

File: vault/share/plugins/vault/30_generic_methods/variable_from_vault.cf [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+      "config_defined" expression => isvariable("vault.config");
+
+  methods:
+    pass1::
+      "Auth" usebundle => vault_get_token("userpass"),
+        ifvarclass => not(isvariable("vault.auth_token"));
+
+    pass3.variable_defined::
+      "success" usebundle => _classes_repaired("${old_class_prefix}");
+      "success" usebundle => _classes_repaired("${class_prefix}");
```
Comment (ncharles) [inline vault/share/plugins/vault/30_generic_methods/variable_from_vault.cf]: for the sanity of the person doing maintenance in the futur, could you change the promiser to "repair" ?

**Comment by VictorQrt** [comment]
Commit modified

File: vault/share/plugins/vault/20_cfe_basics/vault_lib.cf [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#####################################################################################
+
+# Fetches the configuration to be used for this run.
+# The config is located at "config_path". It is a JSON file containing two fields : a server_addr and an auth dictionary.
```
Comment (ncharles) [inline vault/share/plugins/vault/20_cfe_basics/vault_lib.cf]: can you paste the sample, so that it'll be easier for users ?

**Comment by VictorQrt** [comment]
Commit modified

**Comment by VictorQrt** [comment]
Commit modified

**Comment by peckpeck** [comment]
OK, merging this PR

**Pull request merged**